### ISSUE TITLE
TF Serving padding issue for Transformer

### ISSUE THREAD

**Issue by oleg-yaroshevskiy** [tensorflow/tensor2tensor#1489]
Hi, 
For long time I'm trying to solve problem with tf serving for transformer nmt model inference. So let's assume I create simple signature: 

```
inputs = tf.placeholder(dtype=tf.int64, shape=[None, None], name="input_logits")
features = {"inputs": standardize_shapes(inputs)}

        return tf.estimator.export.ServingInputReceiver(
            features=features, receiver_tensors=inputs)

```
and everything works well.
Then I want to enable batching and due to different sequence lengths I set 
`pad_variable_length_inputs: true ` in batching.config file.
Thats where the problem starts. Serving returns garbage for short sequences f.e:
```
first (1, 18) {'instances': [{'inputs': [336, 201, 506, 26, 1902, 2339, 6677, 13748, 14, 55, 11864, 3258, 7380, 1368, 2770, 26172, 662, 1]}]} enfr {
    "predictions": [
        {
            "outputs": [367, 169, 1306, 2, 37, 10043, 301, 19606, 4, 19005, 151, 85, 15888, 11427, 20653, 5131, 21936, 16, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
            "scores": -9.25369
        }
    ]
}

second (1, 3) {'instances': [{'inputs': [1544, 657, 1]}]} enfr {
    "predictions": [
        {
            "outputs": [2865, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 1141, 2865, 2865, 2865, 2865, 2865, 2865, 2865, 2865, 2865, 2865, 2865, 2865, 2865, 1],
            "scores": -28.8886
        }
    ]
}
```

as we can see a longer one was predicted well and for short one it messed up with repeated tokens. This can't be reproduced for single element inference or out of tf serving env.

Example after decoding:
```
['Made from steel']->['MadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeM
adeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMadeMa
deMadeMadeMadeMadeMademade.']
```

Any idea? Is it related to tf-serving padding? Or not the best input signature?
Give me a clue,
thanks.

### Environment information
tf serving 1.13.0
tensor2tensor==1.6.6

**Comment by oleg-yaroshevskiy** [comment]
Can't be reproduced without `pad_variable_length_inputs: true` so I've created a ticket under serving repo. Will close this one asap

**Comment by oleg-yaroshevskiy** [comment]
I've located the problem in [batching_util.cc](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/batching/batching_util.cc)
```
    T pad_value(input.flat<T>()(0));  // using existing values in padding
    output->tensor<T, num_dims>() = inputs.pad(padding, pad_value);
```
Looks like we pad a tensor with first value of flatten tensor (or?) and not with a zero and that probably might work for image processing but not good for sequence batching.
___
So that won't work for t2t NMT task as we want to pad tensors with zeros. Closing it here.

**Comment by BorisPolonsky** [comment]
Hello, I would like to know how you reproduced this issue. I have my batch config like this:
```
max_batch_size { value: 128 }
batch_timeout_micros { value: 0 }
max_enqueued_batches { value: 1000000 }
num_batch_threads { value: 8 }
pad_variable_length_inputs: true
```
But the tensorflow/serving API won't accept inputs with variable length in a single batch. e.g.
```
{
    'instances': [{'inputs': [1,2,3]}, {'inputs': [1,2,3, 4]}]}
    ]
}
```
